src/main/native/GCM.c

```diff
@@ -41,12 +41,41 @@ typedef struct {
 #define THREAD_LOCAL __thread
 #endif
 
-int  GCM_InitForUpdateDecrypt_core ();
-int  GCM_UpdForUpdateDecrypt_core ();
-int  GCM_FinalForUpdateDecrypt_core ();
-int  GCM_InitForUpdateEncrypt_core ();
-int  GCM_UpdForUpdateEncrypt_core ();
-int  GCM_FinalForUpdateEncrypt_core ();
+int GCM_InitForUpdateDecrypt_core(JNIEnv *env, ICC_CTX* ockCtx, ICC_AES_GCM_CTX* gcmCtx,
+    unsigned char * key, int keyLen,
+    unsigned char * iv,  int ivLen,
+    unsigned char * aad, int aadLen, unsigned long *updateOutlen);
+int GCM_UpdForUpdateDecrypt_core(JNIEnv *env, ICC_CTX* ockCtx, ICC_AES_GCM_CTX* gcmCtx,
+    unsigned char * data, int dataOffset, int dataLen,
+    unsigned char * out, int outOffset,
+    unsigned long *updateOutlen);
+int GCM_FinalForUpdateDecrypt_core(JNIEnv *env, ICC_CTX* ockCtx, ICC_AES_GCM_CTX* gcmCtx,
+    unsigned char *data, int dataOffset, int dataLen,
+    unsigned char * out, int outOffset, 
+    int tagLen, unsigned long updateOutlen);
+int GCM_InitForUpdateEncrypt_core(JNIEnv* env, ICC_CTX* ockCtx, ICC_AES_GCM_CTX* gcmCtx,
+    unsigned char* key   , int keyLen,
+    unsigned char* iv    , int ivLen,
+    unsigned char* aad   , int aadLen,
+    unsigned long *updateOutlen);
+int GCM_UpdForUpdateEncrypt_core(JNIEnv* env, ICC_CTX* ockCtx, ICC_AES_GCM_CTX* gcmCtx,
+    unsigned char* data , int dataLen, int dataOffset,
+    unsigned char* out, int outOffset, unsigned long* updateOutlen);
+int GCM_FinalForUpdateEncrypt_core(JNIEnv* env, ICC_CTX* ockCtx, ICC_AES_GCM_CTX* gcmCtx,
+    unsigned char* tag   , int tagLen,
+    unsigned char *dataText, int dataOffset, int dataLen, 
+    unsigned char* out, int outOffset, unsigned long updateOutlen);
+
+typedef size_t UDATA;
+typedef signed char* arr;
+
+// Pointers of functions that are only available on some hardware (might be null)
+typedef void (*ECB_FuncPtr)(signed char *, signed char *, UDATA *, signed char*, UDATA *);
```

Comment:
Those seem exactly the same as in `CCM`. Should we extract them into a header file or something?

Reply:
I consolidated the pointer defintions in the file `zHardwareFunctions.h`. In this file there are now defintions used in multiple places named `ECB_FuncPtr`, `GHASH_FuncPtr`, `zS390_FuncPtr`, and `KMC_FuncPtr`.